cbmc_viewer/byteopt.py

```diff
@@ -0,0 +1,143 @@
+            self.__dict__, VALID_BYTE_OP_SUMMARY
+        )
+
+    def dump(self, filename=None, outdir='.'):
+        """Write the byte op summary to a file rendered as html."""
+
+        util.dump(self, filename or "byteop.html", outdir)
+
+################################################################
+# Json key tags to access elements in cbmc json output
+
+JSON_EXTRACT_KEY = 'byteExtractStats'
+JSON_EXTRACT_LIST_KEY = 'byteExtractList'
+JSON_EXTRACT_COUNT_KEY = 'numOfExtracts'
+JSON_UPDATE_KEY = 'byteUpdateStats'
+JSON_UPDATE_LIST_KEY = 'byteUpdateList'
+JSON_UPDATE_COUNT_KEY = 'numOfUpdates'
+JSON_BYTE_OP_KEY = 'byteOpsStats'
+JSON_SOURCE_LOC_KEY = 'sourceLocation'
+
+################################################################
+# Utility functions to generate byteop summary
+
+def extract_byte_op_data(json_data):
+    for item in json_data:
+        if JSON_BYTE_OP_KEY in item.keys():
+            return item[JSON_BYTE_OP_KEY]
+    return None
+
+def remove_ssa_expr(json_data):
+    JSON_REMOVE_KEY = 'ssaExpr'
+
+    json_extract_list = json_data.get(JSON_EXTRACT_KEY).get(JSON_EXTRACT_LIST_KEY)
+    json_update_list = json_data.get(JSON_UPDATE_KEY).get(JSON_UPDATE_LIST_KEY)
+
+    for item in json_extract_list:
+        item.pop(JSON_REMOVE_KEY)
+    for item in json_update_list:
+        item.pop(JSON_REMOVE_KEY)
+
+def get_src_location(json_data, root):
+    json_extract_list = json_data.get(JSON_EXTRACT_KEY).get(JSON_EXTRACT_LIST_KEY)
+    json_update_list = json_data.get(JSON_UPDATE_KEY).get(JSON_UPDATE_LIST_KEY)
+
+    root = srcloct.abspath(root)
+    for item in json_extract_list:
+        item[JSON_SOURCE_LOC_KEY] = srcloct.json_srcloc(item[JSON_SOURCE_LOC_KEY], root)
+    for item in json_update_list:
+        item[JSON_SOURCE_LOC_KEY] = srcloct.json_srcloc(item[JSON_SOURCE_LOC_KEY], root)
+
+def get_byte_op_metrics(json_file, root):
+    summary = {}
+
+    json_data = parse.parse_json_file(json_file)
+    for item in json_data:
+        if JSON_BYTE_OP_KEY in item.keys():
+            byteop = item[JSON_BYTE_OP_KEY]
+            remove_ssa_expr(byteop)
+            get_src_location(byteop, root)
+
+            summary['byteExtractList'] = byteop[JSON_EXTRACT_KEY][JSON_EXTRACT_LIST_KEY]
+            summary['numOfExtracts'] = byteop[JSON_EXTRACT_KEY][JSON_EXTRACT_COUNT_KEY]
+            summary['byteUpdateList'] = byteop[JSON_UPDATE_KEY][JSON_UPDATE_LIST_KEY]
+            summary['numOfUpdates'] = byteop[JSON_UPDATE_KEY][JSON_UPDATE_COUNT_KEY]
+            break
+
+    if not summary:
+        summary = {
+            'byteExtractList': [],
+            'numOfExtracts': 0,
+            'byteUpdateList': [],
+            'numOfUpdates': 0
+        }
+
+    return summary
+
+def fail(msg):
+    """Log failure and raise exception."""
+
+    logging.info(msg)
+    raise UserWarning(msg)
+
+def do_make_byteop(cbmc_byteop, srcdir):
+    if srcdir and cbmc_byteop:
```

Comment:
This is how its being done throughput viewer, although I'm not sure why.
@markrtuttle will be able to tell us more. I assumed this was an additional check cause srcdir and cbmc_byteop are got from command line and if they are empty it would raise an exception during argument parsing.

Reply:
This do_make_byteop has functionality that is very different from other object.  In all other objects, the return value is an object like a ByteOpSummary object.  In contrast, this function is used by the init method of ByteOpSummary.

The purpose of the do_make_TYPE functions is to have a uniform way for creating the TYPE object in the make_TYPE script and in the main body of viewer.